Fix Xf86videointel driver immediate crash on NixOS (Mesa Patch Backport)

tjkeller-xyz opened the pull request:
<!--
The 
-->

This PR adds a patch that fixes an upstream issue in Mesa which breaks the xf86videointel driver on NixOS. I am going to submit this patch to Mesa upstream as well, but it would be very nice to have this backported to 24.11 :)

The issue was introduced in Mesa 24.2.2 in [this patch](https://gitlab.freedesktop.org/mesa/mesa/-/commit/06d417af80bc1f171cadc338e63a7aa75c877754.diff). Not checking for an invalid config causes X to crash immediately when launching with the xf86videointel driver.

The xf86videointel driver is **required** on i915 gpus, the modesetting driver simply crashes on startup.

This also re-adds the xf86videointel driver to xorg's `overrides.conf`, and updates it to the latest version (2024-05-06). Currently it throws an error when used on 24.11 and unstable.

Please see [this thread](https://discourse.nixos.org/t/xinit-segfault-unw-get-proc-name-failed-no-unwind-info-found-10-intel-gma-i915-modesetting-does-not-work-have-to-use-xf86videointel/59060/11) on NixOS Discourse for additional context.

Thanks!

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

tjk918 commented:
> That sounds very wrong. How does the modesetting driver crash?

This is a known issue on i915 gpus (which at this point are 15+ years old). You can see the backtrace from startx [here](https://discourse.nixos.org/t/xinit-segfault-unw-get-proc-name-failed-no-unwind-info-found-10-intel-gma-i915-modesetting-does-not-work-have-to-use-xf86videointel/59060) on the NixOS Discourse thread.

Also [here](https://github.com/NixOS/nixpkgs/issues/342763#issuecomment-2535831051)

tjk918 commented:
Im not sure. I am currently submitting a merge request on mesas gitlab. The issue had only been observed on NixOS as far as I can tell, and was much reported here until the xf86videointel driver was simply removed for causing issues. There really wasnt much demand to fix it since it people using newer intel gpus can just use modesetting instead.

 Other distributions work just fine with the intel driver even with newer versions of Mesa. But the patch that was added in 24.2.2 (linked above) is clearly the culprit and seems to cause the crash because the check for a 'null config' was removed (almost seemingly by accident).

vcunat commented:
[Merged upstream](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/33336#note_0c9e89d8d9ea3dafb3426b737be6909ac79fd640), so we fetch from the merged commit instead of putting the patch into nixpkgs, right?

tjk918 commented:
> Nit: [merged upstream](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/33336#note_0c9e89d8d9ea3dafb3426b737be6909ac79fd640), so we fetch from the merged commit instead of putting the patch into nixpkgs, right?

I believe so. I will update the pull request to do just that!

Commit:
mesa: backport dril patch that fixes xf86videointel

Commit:
re-add and update xf86videointel driver

tjk918 commented:
Fetchpatch is now used to grab the upstream patch, and I did a forced update so that the patch wont be included in the git tree

vcunat reviewed (approved):
Looks good.  I haven't tested anything.

tjkeller-xyz committed:
Add fetchpatch dependency

tjk918 commented:
@K900 @rnhmjoj Sorry to ping you both but I am hoping to get this reviewed and merged soon :). Not in any particular rush, just seems like the PR is kinda dead at the moment and I would like to get it back on the radar in case it has been forgotten. If there is anything else I should need to take care of first, please let me know. Thanks

rnhmjoj commented:
Sorry for the delayed reply: I tested your patch by backporting it to 24.11 and can confirm the driver resumed working.

Can you update the NixOS manual section on Intel graphics and add a release note that says the driver is back?

Commit:
nixos/doc: update the status of Intel Graphics with fixed intel driver

xf86-video-intel is now working again due to a bug fix in Mesa. So this
re-introduces the section on the intel driver and screen-tearing while
discouraging its use generally.

tjkeller-xyz commented:
@rnhmjoj No problem at all. I added a commit that removes from the release notes the section about the broken driver, and also re-introduces the info about the intel driver back into the manual. Please lmk if there are any changes you would like me to make. Thanks!

rnhmjoj commented:
I'm not sure about the release notes: we should probably not change them retroactively, but if we backport the fix then they become incorrect... What's certainly safe to do is to mentione that the driver is fixed in the upcoming 25.05.

Commit:
Mention xf86-video-intel fix in 25.05 release notes and undo change to 24.11 release notes

rnhmjoj reviewed (approved):
It looks all good. Thank you for the excellent work!

tricktron commented:
I manually tested the patch and this will also fix the X crash experienced in VMware guest: https://github.com/NixOS/nixpkgs/issues/282155 and makes it possible to use VMware with 3D acceleration again. Thank you all for fixing this!

vcunat commented:
So I reverted that merge.  No point in this until the driver builds :-)

tjkeller-xyz commented:
What is preventing the driver from building? The derivarion is the same as it was prior to the driver being removed, it should build just fine.

tjkeller-xyz commented:
I successfully built this just now on nixos (24.11) using this in my config:
```
environment.systemPackages = with pkgs; [
	xorg.xf86videointel
];

nixpkgs.overlays = [
	(self: super: {
		xorg = (super.xorg or { }) // {
		xf86videointel = super.xorg.xf86videointel.overrideAttrs (attrs: {
		  # the update script only works with released tarballs :-/
		  name = "xf86-video-intel-2024-05-06";
		  src = pkgs.fetchFromGitLab {
		    domain = "gitlab.freedesktop.org";
		    group = "xorg";
		    owner = "driver";
		    repo = "xf86-video-intel";
		    rev = "ce811e78882d9f31636351dfe65351f4ded52c74";
		    sha256 = "sha256-PKCxFHMwxgbew0gkxNBKiezWuqlFG6bWLkmtUNyoF8Q=";
		  };
		  buildInputs = attrs.buildInputs ++ [ pkgs.xorg.libXScrnSaver pkgs.xorg.libXv pkgs.xorg.pixman pkgs.xorg.utilmacros ];
		  nativeBuildInputs = attrs.nativeBuildInputs ++ [pkgs.autoreconfHook ];
		  configureFlags = [ "--with-default-dri=3" "--enable-tools" ];
		  patches = [ ./use_crocus_and_iris.patch ];

		  meta = attrs.meta // {
		    platforms = ["i686-linux" "x86_64-linux"];
		  };
		});
		};
	})
];
```

Which is the exact same derivation that was included in my PR.

Can you send me logs of the build process failing? Thanks. I suppose I will have to create a new PR for this to be merged again once it is fixed

K900 commented:
We'll have Mesa 25.0.1 in a week, which includes the patch.

vcunat commented:
On the commit on master (2092d8064405) it won't build for me.  Easy for you to retry the same, right?

vcunat commented:
Anyway, log for reference:
<details><pre>
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/mca3plkvl1f0s2f1ajay7x8vics3yl23-source
source root is source
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
applying patch /nix/store/2fzz1gydljpzr00kai0q2lysnaw3k6cp-use_crocus_and_iris.patch
patching file src/uxa/intel_dri.c
patching file src/sna/sna_dri2.c
Hunk #1 succeeded at 3709 (offset 2 lines).
Running phase: autoreconfPhase
@nix { "action": "setPhase", "phase": "autoreconfPhase" }
autoreconf: export WARNINGS=
autoreconf: Entering directory '.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal --force -I m4
configure.ac:38: error: must install xorg-macros 1.8 or later before running autoconf/autogen.
  Hint: either install from source, git://anongit.freedesktop.org/xorg/util/macros or,
  depending on you distribution, try package 'xutils-dev' or 'xorg-x11-util-macros'
configure.ac:38: the top level
autom4te: error: /nix/store/v9ck7wksinqk161hgclbf16g3c8lc46z-gnum4-1.4.19/bin/m4 failed with exit status: 1
aclocal: error: /nix/store/m9qpqr7m6ik1szwlrnk2sjw7x7k0rdbg-autoconf-2.72/bin/autom4te failed with exit status: 1
autoreconf: error: aclocal failed with exit status: 1

rnhmjoj commented:
The fix seems to be:
```patch
diff --git a/pkgs/servers/x11/xorg/overrides.nix b/pkgs/servers/x11/xorg/overrides.nix
index 0d3b3a5c2dc9..31437cd944a8 100644
--- a/pkgs/servers/x11/xorg/overrides.nix
+++ b/pkgs/servers/x11/xorg/overrides.nix
@@ -1020,8 +1020,8 @@ self: super:
       rev = "ce811e78882d9f31636351dfe65351f4ded52c74";
       sha256 = "sha256-PKCxFHMwxgbew0gkxNBKiezWuqlFG6bWLkmtUNyoF8Q=";
     };
-    buildInputs = attrs.buildInputs ++ [ xorg.libXScrnSaver xorg.libXv xorg.pixman xorg.utilmacros ];
-    nativeBuildInputs = attrs.nativeBuildInputs ++ [autoreconfHook ];
+    buildInputs = attrs.buildInputs ++ [ xorg.libXScrnSaver xorg.libXv xorg.pixman ];
+    nativeBuildInputs = attrs.nativeBuildInputs ++ [ autoreconfHook xorg.utilmacros xorg.xorgserver ];
     configureFlags = [ "--with-default-dri=3" "--enable-tools" ];
     patches = [ ./use_crocus_and_iris.patch ];
 
```

tjkeller-xyz commented:
Awesome thank you! I added the fix to my branch. Should I create a new PR?

vcunat commented:
Created as https://github.com/NixOS/nixpkgs/pull/385760
